typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,726 @@
+    if (fn(arr[i])) {
+      newArray.push(arr[i]);
+    }
+  }
+  return newArray; // or just simple return arr.filter(fn);
+};
+
+const f = (x: string): ((s: number) => boolean) => {
+  return (s: number) => {
+    return s.toString() === x;
+  };
+};
+
+f("5")(5);
+```
+
+As you may see in the last example, we can call function `f` and return function from it, applying in for the new argument.
+So, yes, we can implement some king of currying using it.
+
+```typescript
+type addT = (_: number) => (_: number) => number;
+const add: addT = (l) => (r) => l + r;
+
+type foldArrT = (f: (_: number) => (_: number) => number) => (_: number) => (_: number[]) => number;
+const foldArr: foldArrT = (f) => (z) => (arr) => {
+  let m = z;
+  arr.forEach((elem) => {
+    m = f(m)(elem);
+  });
+  return m;
+};
+
+foldArr(add)(0)([1, 2, 3, 4, 5]);
+```
+
+Unfortunately syntax is weird as you need to always use named parameters.
+But it is still a possible functionality, and you can simply use it in your code.
+
+### Polymorphism
+
+There are different polymorphisms, but we will say only about thee of them: Parametric, Ad hoc, and Row.
+Let's start from the first on.
+Parametric polymorphism allows us to write abstract function or data which will not depend on its type.
+In TypeScript, it is called generics.
+
+```typescript
+const arrLength: <T>(_: T[]) => number = (arr) => arr.length;
+
+arrLength<number>([1, 2, 3]);
+arrLength<string>(["1", "2", "3"]);
+
+
+type NonEmpty<T> = {
+  head: T;
+  tail: T[];
+};
+```
+
+Ad hoc polymorphism allows you to implement abstract functions, which logic will be different with different types.
+
+```typescript
+interface Eq<T> {
+  equal: (f: T, s: T) => boolean;
+}
+
+class IntEq implements Eq<number> {
+  equal(f: number, s: number) {
+    return f === s;
+  }
+}
+
+class IntArrEq implements Eq<number[]> {
+  equal(f: number[], s: number[]) {
+    return f.filter(item => s.indexOf(item) < 0).length === 0;
+  }
+}
+
+const lookup = <T, K extends Eq<T>, V>(cmp: K, k: T, mp: [T, V][]): V | undefined => {
+  let result: V | undefined;
+  mp.forEach(([kk, v]) => {
+    if (cmp.equal(k, kk)) {
+      result = v;
+    }
+  });
+  return result;
+};
+
+lookup(new IntEq(), 1, [[1, "2"], [2, "1"]]);
+
+lookup(new IntArrEq(), [1, 2, 3], [[[1, 3], "1"], [[3, 2], "2"], [[2, 1, 3], "3"]]);
+```
+
+One more polymorphism is row.
+It allows you to abstract by the record fields and allow only those of them which have such fields.
+Unfortunately TypeScript type compatibility is based on structural subtyping, and it is not a row polymorphism.
+Despite this we can use intersection types to emulate this.
+But you should be careful with using it because subtyping cause information loss as if you have value `a: T` and `b: T & {x: number}` you will be able to do `a = b` and cut off `b` to `T`.
+
+```typescript
+type fnT = <T>(v: T & { x: number }) => T & { x: number }
+// You should fully describe all types to not lose information.
```

Comment:
I think in `Elm` you just can't return `x` from the function with type `{x : Int | t} -> t` \
I mean you can't write such thing
```elm
type alias A =
  { x : Int
  , y : Int
  }

type alias B t = {t | x : Int}
  
f : B t -> t
f a = a
```

But in TS it is possible
```typescript
type A = { x: number; y: number };

type B<T> = T & { x: number };

const f: <T>(a: B<T>) => T = (a) => a;
```
So, in case of `Elm` yes, we want to remove field `x` as if we will not remove it compiler will say about it to us. But in case of TS it will say nothing.

Reply:
>remove a field

Moreover it will not remove it at all but only on type level.
```
type A = {};
type B = A & { x: number };

let a: A = {};
const b: B = { x: 1 };
a = b;
console.log(JSON.stringify(a)); // {"x":1}
```

Reply:
So, I think `const f: <T>(a: B<T>) => T = (a) => a;` cause bugs and you should `fully describe all types to not lose information`

Reply:
>fully describe all types

I think it is the biggest problem with TS when you try to go a little bit deeper. Compiler will help you only if you will help him